mysql/token_provider.go

```diff
@@ -150,9 +151,18 @@ func (ct *cachedToken) updateToken(ctx context.Context, c *mysql.Config) error {
 	); err != nil {
 		return err
 	}
+	
+	// Parse expiry before updating the cached state to avoid inconsistency
+	newExpires, err := expiry(newToken)
+	if err != nil {
+		return err
+	}
+	
+	// Only update the cached state if both token generation and expiry parsing succeed
+	ct.token = newToken
```

Comment:
I don't care about preserving the old token, but the time we call `update` we already have determined that the old token is stale.